File: typescript/gql-codegen/src/schema/client/operation.ts
import { z } from 'zod/v4';
import { fragmentSpecSchema } from './fragment.js';
import { inputFieldSchema } from '../shared.js';

export const operationSchema = z.object({
    name: z.string(),
    type: z.enum(['MUTATION', 'QUERY', 'SUBSCRIPTION']),
    parameters: z.record(z.string(), inputFieldSchema),
    fragmentSpec: fragmentSpecSchema,
    sourceText: z.string(),
    hash: z.string()
})

File: rust/gqlcodegen/src/schema/client/operation.rs
use std::collections::HashMap;

use super::fragment::FragmentSpec;
use crate::schema::shared::InputField;

#[derive(Debug, serde::Deserialize)]
pub enum OperationType {
    #[serde(rename(deserialize = "MUTATION"))]
    Mutation,
    #[serde(rename(deserialize = "QUERY"))]
    Query,
    #[serde(rename(deserialize = "SUBSCRIPTION"))]
    Subscription,
}

#[derive(Debug, serde::Deserialize)]
pub struct Operation {
    name: String,
    #[serde(rename(deserialize = "type"))]
    optype: OperationType,
    parameters: HashMap<String, InputField>,
    #[serde(rename(deserialize = "fragmentSpec"))]
    fragment_spec: FragmentSpec,
    #[serde(rename(deserialize = "sourceText"))]
    source_text: String
}
